Handle nuget packages with per-platform runtimes

bucurb opened the pull request:
This PR fixes #221 
- Added a `PlatformDefinition` class as a replacement for the  `PlatformEnables` in `NativeLibraries`.
- Unified the meta files for the lib and native dlls.
- Added Android and WebGL to the dll meta files.
- Dlls in nuget packages that have `runtimes` folders compatible with the framework that's getting pulled into UPM are also written into the UPM. In this  case, the dlls in the `lib` folder continue to be written into the UPM, but only for the platforms / architectures not covered by the `runtimes` dlls.

bdovaz commented on src/UnityNuGet/RegistryEntry.cs:
```diff
@@ -15,6 +15,9 @@ public class RegistryEntry
         public bool Listed { get; set; }
 
+        [JsonProperty("scope")]
+        public string? UnityScope { get; set; }
```
I understand this has to do with https://github.com/xoofx/UnityNuGet/issues/216 so I think you have mixed this PR with a different issue, no?

Also I already argued why we can't/shouldn't introduce this behavior.

bucurb replied:
I removed that, sorry

bdovaz commented on src/UnityNuGet/PlatformDefinition.cs:
```diff
@@ -0,0 +1,306 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using static System.Runtime.InteropServices.JavaScript.JSType;
```
Remove this using please

bucurb replied:
done.

bdovaz commented:
First and foremost, you have to admit it, very good job! 

For the part of the goals you had to do a lot of reverse engineering because Unity has absolutely nothing documented about `platformData`. In fact, there was a request but it came to nothing https://github.com/xoofx/UnityNuGet/issues/40#issuecomment-938708659 for this reason.

And for the NuGet part I also see that you have done a lot of research, thank you very much.

Comments:

1. I am concerned that there are no tests for absolutely nothing you have added. There are a lot of substring, split that doesn't even have comments in the code (or a Microsoft docs url) to see where that definition/logic comes from. I would be more confident if you added tests with inputs from specific cases you have seen to see that the result is as expected. Manipulating the meta files has implications of breaking an assembly import if it is not well constructed.

2. Why didn't you include iOS? It is supposed to be supported, isn't it? https://learn.microsoft.com/en-us/dotnet/core/rid-catalog#ios-rids

xoofx commented:
Don't forget that if the content of packages are changed (especially for old packages), the version of the registry should be bumped in order to regenerate them:

https://github.com/xoofx/UnityNuGet/blob/614b8d587800c2913eeadd1fbc7f31bf39b01689/src/UnityNuGet/RegistryCache.cs#L35-L36

bdovaz commented:
@xoofx I already have it in mind (but thanks), I had to do it in `0.29.0` for this PR https://github.com/xoofx/UnityNuGet/pull/198

bucurb commented:
@xoofx done, bumped the version to 1.6.0.
@bdovaz I've added comments to the code and unit tests for everything I could think of. We now unit test the `PlatformDefinition` configurations& tree, as well as the file paths constructed from them. I've added a test that just fetches `System.Security.Cryptography.ProtectedData` version 6.0.0 and ensures it has runtime overrides for Windows that completely replace the lib files. Finally, I've added Unity meta tests that ensure only the expected platform bits are enabled. If you've anything else in mind that can be tested just let me know and can add.

iOS support is also in.
